### PR TITLE
[CCR] Added write buffer size limit

### LINKED ISSUE #34705: [CCR] Limit the write buffer by memory rather than op count
Currently we stop sending read requests if the number of operation waiting to sent as a write bulk is higher than a given number. That was the simplest to implement but it makes it harder to reason about a good value - the ultimate goal of setting a limit is to protect against memory consumption. I think we should move to a byte sized based limit.

/cc @jasontedor @martijnvg

### PR DESCRIPTION
**Pull request by martijnvg**
This limit is based on the size in bytes of the operations in the write buffer. If this limit is exceeded then no more read operations will be coordinated until the size in bytes of the write buffer has dropped below the configured write buffer size limit.

Renamed existing `max_write_buffer_size` to `max_write_buffer_count` to indicate that limit is count based.

Closes #34705

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java [hunk 1]
```diff
@@ -327,7 +327,7 @@ private void followLeaderIndex(String autoFollowPattenName,
             followRequest.setMaxConcurrentReadBatches(pattern.getMaxConcurrentReadBatches());
             followRequest.setMaxBatchSize(pattern.getMaxBatchSize());
             followRequest.setMaxConcurrentWriteBatches(pattern.getMaxConcurrentWriteBatches());
-            followRequest.setMaxWriteBufferSize(pattern.getMaxWriteBufferSize());
+            followRequest.setMaxWriteBufferCount(pattern.getMaxWriteBufferCount());
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java]: We forget passing "maxWriteBufferSize" parameter.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java [hunk 2]
```diff
@@ -85,6 +85,7 @@
     private long numberOfOperationsIndexed = 0;
     private long lastFetchTime = -1;
     private final Queue<Translog.Operation> buffer = new PriorityQueue<>(Comparator.comparing(Translog.Operation::seqNo));
+    private long bufferSize = 0;
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: Should it be bufferSizeInBytes?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java [hunk 3]
```diff
@@ -208,6 +213,8 @@ private synchronized void coordinateWrites() {
                     break;
                 }
             }
+            long opsSize = ops.stream().mapToLong(Translog.Operation::estimateSize).sum();
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: Can we move this into the loop to avoid the loop.

**Review by dnhatn — COMMENTED** [review]
This looks great. I left some minor comments.

I noticed that the current implementation (not this PR) might fill the write buffer more than the limit. When we send a read-request, we don't limit the request count and size to the vacant slots of the write buffer. If the buffer has one byte (or one count) left, we still issue a full read-request. We can fix this in a follow-up if we feel we should.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java (hunk 1 above)
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java]: whoops

**Comment by martijnvg** [comment]
@dnhatn Thanks for reviewing!

> I noticed that the current implementation (not this PR) might fill the write buffer more than the limit. When we send a read-request, we don't limit the request count and size to the vacant slots of the write buffer. If the buffer has one byte (or one count) left, we still issue a full read-request.

Yes, that is what it is currently doing. I don't see a real problem in the way the limit is currently enforced.

**Comment by bleskes** [comment]
> Yes, that is what it is currently doing. I don't see a real problem in the way the limit is currently enforced.

+1. The main rationale here was to keep things simple and treat the limit as a soft limit that is used to cause back pressure. No more.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java [hunk 4]
```diff
@@ -455,6 +466,7 @@ public synchronized ShardFollowNodeTaskStatus getStatus() {
                 numConcurrentReads,
                 numConcurrentWrites,
                 buffer.size(),
+            bufferSizeInBytes,
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: nit: indentation

**Review by dnhatn — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Makes sense. Thanks @martijnvg and @bleskes for explaining.

**Comment by martijnvg** [comment]
Note that this PR also changes the default for `max_write_buffer_count` from `10240` to unlimited:
https://github.com/elastic/elasticsearch/pull/34797/commits/9d642b568b184eeff08812ef9f295c2c363ee675

The reason behind this that a finding a default for this parameter is difficult and because this pr add `max_write_buffer_size`, that will make sure that the write buffer does not accumulate too many write operations.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**